netket/utils/group/permutation_group.py

```diff
@@ -0,0 +1,203 @@
+
+    def __matmul__(self, other) -> "PermutationGroup":
+        if not isinstance(other, PermutationGroup):
+            raise ValueError(
+                "Incompatible groups (`PermutationGroup` and something else)"
+            )
+        elif self.degree != other.degree:
+            raise ValueError(
+                "Incompatible groups (`PermutationGroup`s of different degree)"
+            )
+
+        return PermutationGroup(super().__matmul__(other).elems, self.degree)
+
+    def _canonical(self, x: Element) -> Array:
+        if isinstance(x, Identity):
+            return np.arange(self.degree, dtype=int)
+        elif isinstance(x, Permutation):
+            return np.asarray(x.permutation)
+        else:
+            raise ValueError(
+                "`PermutationGroup` only supports `Identity` and `Permutation` elements"
+            )
+
+    def to_array(self) -> Array:
+        """
+        Convert the abstract group operations to an array of permutation indices,
+        such that the `i`-th row contains the indices corresponding to the `i`-th group
+        element. That is, `self.to_array()[i, j]` is :math:`g_i^{-1}(j)`) and
+        (for :code:`G = self`)::
+            V = np.arange(G.degree)
+            assert np.all(G(V) == V[..., G.to_array()])
+        """
+        return self._canonical_array()
+
+    def __array__(self, dtype=None) -> Array:
+        return np.asarray(self.to_array(), dtype=dtype)
+
+    def remove_duplicates(self, *, return_inverse=False) -> "PermutationGroup":
+        """
+        Returns a new :code:`PermutationGroup` with duplicate elements (that is, elements which
+        represent identical permutations) removed.
+
+        Arguments:
+            return_inverse: If True, also return indices to reconstruct the original
+                group from the result.
+
+        Returns:
+            group: the permutation group with duplicate elements removed.
+            return_inverse: Indices to reconstruct the original group from the result.
+                Only returned if `return_inverse` is True.
+        """
+        if return_inverse:
+            group, inverse = super().remove_duplicates(return_inverse=True)
+        else:
+            group = super().remove_duplicates(return_inverse=False)
+
+        pgroup = PermutationGroup(group.elems, self.degree)
+
+        if return_inverse:
+            return pgroup, inverse
+        else:
+            return pgroup
+
+    @struct.property_cached
+    def inverse(self) -> Array:
+        try:
+            lookup = self._canonical_lookup()
+            inverses = []
+            for perm in self.to_array():
+                invperm = np.argsort(perm)
+                inverses.append(lookup[HashableArray(invperm)])
+
+            return np.asarray(inverses, dtype=int)
+        except KeyError:
+            raise KeyError(
+                "PermutationGroup does not contain the inverse of all elements"
+            )
```

Comment:
This should be either something generic (`RuntimeError`) or a custom error, `KeyError` doesn't make much sense if no key is supplied.